### PR TITLE
Node: `UrlObject`: omit `path`

### PR DESCRIPTION
**Pull request by OliverJAsh**
Incorrect:

```ts
url.format({ path: '/foo?a=b' })
// => ''
```

Correct:

```ts
url.format({ pathname: '/foo', search: '?a=b' })
// => '/foo?a=b'
```

---

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://devdocs.io/node/url#url_url_format_urlobject
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

### CODE REVIEW

**Review by islishude — CHANGES_REQUESTED** [review]
see https://nodejs.org/dist/latest-v12.x/docs/api/url.html#url_urlobject_path

**Comment by OliverJAsh** [comment]
@islishude IIUC, `path` exists on the _output_ of `url.parse`, but it is never used as a parameter inside of `url.format`. In that case, `search` and `pathname` must be used instead.

The `UrlObject` type in this project reflects the latter part only, as indicated by the comment immediately above the type.

For that reason I think `path` should be considered an invalid parameter and thus removed from the `UrlObject` type?

If `UrlObject` is used in more places than just the input of `url.format`, then we need a separate type.

**Comment by islishude** [comment]
url.format accepts [url.URL class](https://nodejs.org/dist/latest-v12.x/docs/api/url.html#url_the_whatwg_url_api) instead of [url. urlObject](https://nodejs.org/dist/latest-v12.x/docs/api/url.html#url_legacy_url_api)

**Comment by islishude** [comment]
FYI 
https://github.com/nodejs/node/blob/525fd9ca7c6146d15b4f73dcdb3854b9c63e8fbb/lib/url.js#L536

**Comment by OliverJAsh** [comment]
That's not what it says here: https://devdocs.io/node/url#url_url_format_urlobject

I think you're talking about a different thing under the same name: https://devdocs.io/node/url#url_url_format_url_options

File: types/node/node-tests.ts [hunk 1]
```diff
@@ -21,6 +21,10 @@ import Module = require("module");
     {
         url.format(url.parse('http://www.example.com/xyz'));
 
+        // TODO: why won't this work?
```
Comment (galkin) [inline types/node/node-tests.ts]: Please remove this comment

File: types/node/url.d.ts [hunk 2]
```diff
@@ -8,7 +8,6 @@ declare module "url" {
         host?: string | null;
         hostname?: string | null;
         href?: string | null;
-        path?: string | null;
```
Comment (galkin) [inline types/node/url.d.ts]: Please add similar change  to node v10

**Review by galkin — CHANGES_REQUESTED** [review]
Thank you for your pull. According my tests and checking documentation https://nodejs.org/dist/latest-v12.x/docs/api/url.html#url_url_format_urlobject
url.format doesn't work with `path`. `pathname` should be used.

File: types/node/test/url.ts [hunk 3]
```diff
@@ -0,0 +1,153 @@
+import assert = require("assert");
```
Comment (OliverJAsh) [inline types/node/test/url.ts]: I had to move the tests here to workaround https://github.com/DefinitelyTyped/DefinitelyTyped/issues/40967

File: types/node/test/url.ts [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+        pathname: 'search',
+        query: { q: "you're a lizard, gary" }
+    });
+
+    const myURL = new url.URL('https://a:b@你好你好?abc#foo');
+    url.format(myURL, { fragment: false, unicode: true, auth: false });
+
+    // `format doesn't work with `path`: use `pathname` and `search` instead
+    // $ExpectError
+    url.format({ path: '/foo' });
```
Comment (OliverJAsh) [inline types/node/test/url.ts]: This is the only new test

**Comment by OliverJAsh** [comment]
@galkin Done

File: types/node/v11/test/url.ts [hunk 5]
```diff
@@ -0,0 +1,5 @@
+import * as url from 'url';
+
+// `format doesn't work with `path`: use `pathname` and `search` instead
+// $ExpectError
+url.format({ path: '/foo' });
```
Comment (galkin) [inline types/node/v11/test/url.ts]: We don't update EOL version. v11 is EOL. Please remove this changes.

File: types/node/v11/tsconfig.json [hunk 6]
```diff
@@ -26,7 +26,8 @@
         "test/util.ts",
         "test/v8.ts",
         "test/worker_threads.ts",
-        "test/zlib.ts"
```
Comment (galkin) [inline types/node/v11/tsconfig.json]: The same.

File: types/node/v11/url.d.ts [hunk 7]
```diff
@@ -7,7 +7,6 @@ declare module "url" {
         host?: string;
         hostname?: string;
         href?: string;
-        path?: string;
```
Comment (galkin) [inline types/node/v11/url.d.ts]: The same.

**Review by galkin — CHANGES_REQUESTED** [review]

**Review by galkin — APPROVED** [review]

**Review by islishude — APPROVED** [review]

**Review by trivikr — APPROVED** [review]

**Comment by DanielRosenwasser** [comment]
The slowdown is likely a fluke - maybe we should re-run tests. Any easy way to do that @sandersn? Also, would be nice if there didn't need to be a new file to work around that linked issue.

**Comment by OliverJAsh** [comment]
@DanielRosenwasser Merged master in so that should force the tests to re-run 🤞

**Review by r3nya — APPROVED** [review]

**Comment by andrewbranch** [comment]
Merge conflicts ☹️

**Pull request merged**